pkg/tfvars/baremetal/baremetal_test.go

```diff
@@ -0,0 +1,164 @@
+// Package baremetal contains bare metal specific Terraform-variable logic.
+package baremetal
+
+import (
+	"encoding/json"
+	"testing"
+
+	"github.com/openshift/installer/pkg/types/baremetal"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestMastersSelectionByRole(t *testing.T) {
+
+	cases := []struct {
+		scenario string
+
+		numControlPlaneReplicas int64
+		libvirtURI              string
+		apiVIP                  string
+		imageCacheIP            string
+		bootstrapOSImage        string
+		externalBridge          string
+		externalMAC             string
+		provisioningBridge      string
+		provisioningMAC         string
+		platformHosts           []*baremetal.Host
+		image                   string
+		ironicUsername          string
+		ironicPassword          string
+		ignition                string
+
+		expectedError string
+		expectedHosts []string
+	}{
+		{
+			scenario:                "filter-workers",
+			numControlPlaneReplicas: 2,
+			platformHosts: platformHosts(
+				host("master-0", "master"),
+				host("master-1", "master"),
+				host("worker-0", "worker"),
+			),
+			expectedHosts: []string{"master-0", "master-1"},
+		},
+		{
+			scenario:                "mixed-order",
+			numControlPlaneReplicas: 1,
+			platformHosts: platformHosts(
+				host("worker-0", "worker"),
+				host("master-0", "master"),
+				host("worker-1", "worker"),
+			),
+			expectedHosts: []string{"master-0"},
+		},
+		{
+			scenario:                "hosts-norole",
+			numControlPlaneReplicas: 2,
+			platformHosts: platformHosts(
+				host("master-0", ""),
+				host("worker-0", "worker"),
+				host("master-1", ""),
+			),
+			expectedHosts: []string{"master-0", "master-1"},
+		},
+		{
+			scenario:                "no-role",
+			numControlPlaneReplicas: 3,
+			platformHosts: platformHosts(
+				host("master-0", ""),
+				host("master-1", ""),
+				host("master-2", ""),
+			),
+			expectedHosts: []string{"master-0", "master-1", "master-2"},
+		},
+		{
+			scenario:                "more-hosts-than-required",
+			numControlPlaneReplicas: 2,
+			platformHosts: platformHosts(
+				host("master-0", "master"),
+				host("master-1", "master"),
+				host("master-2", ""),
+			),
+			expectedHosts: []string{"master-0", "master-1"},
+		},
+		{
+			scenario:                "more-hosts-than-required-norole",
+			numControlPlaneReplicas: 2,
+			platformHosts: platformHosts(
+				host("master-0", ""),
+				host("master-1", ""),
+				host("master-2", ""),
+			),
+			expectedHosts: []string{"master-0", "master-1"},
+		},
+		{
+			scenario:                "more-hosts-than-required-mixed",
+			numControlPlaneReplicas: 2,
+			platformHosts: platformHosts(
+				host("worker-0", "worker"),
+				host("master-0", ""),
+				host("worker-1", "worker"),
+				host("master-1", ""),
+				host("worker-2", "worker"),
+				host("master-2", ""),
+			),
+			expectedHosts: []string{"master-0", "master-1"},
+		},
```

Comment:
We appear to be missing a test case like:

```
	numControlPlaneReplicas: 1,
	platformHosts: platformHosts(
		host("worker-0", ""),
		host("worker-1", ""),
		host("master-0", "master"),
	),
	expectedHosts: []string{"master-0"},
```

Also (spoiler alert!) I don't think it will pass.